api.c

```diff
@@ -905,3 +905,72 @@ int tcmu_emulate_start_stop(struct tcmu_device *dev, uint8_t *cdb,
 
 	return SAM_STAT_GOOD;
 }
+
+int tcmu_emulate_write_verify(struct tcmu_device *dev,
+			      struct tcmulib_cmd *tcmulib_cmd,
+			      ssize_t (*read)(struct tcmu_device *,
+					      struct iovec *, size_t, off_t),
+			      ssize_t (*write)(struct tcmu_device *,
+					       struct iovec *, size_t, off_t),
+			      struct iovec *iovec,
+			      size_t iov_cnt,
+			      off_t offset)
+{
+	struct iovec iov;
+	uint32_t cmp_offset;
+	uint8_t *sense = tcmulib_cmd->sense_buf;
+	uint8_t *cdb = tcmulib_cmd->cdb;
+	size_t length = tcmu_get_xfer_length(cdb) * tcmu_get_dev_block_size(dev);
+	size_t remaining = length;
+	int rc = SAM_STAT_GOOD;
+	int len;
+	int ret;
+
+	while (remaining) {
+		ret = write(dev, iovec, iov_cnt, offset);
+		if (ret < 0) {
+			tcmu_err("writefailed: %m\n");
+			return tcmu_set_sense_data(sense, MEDIUM_ERROR,
+						   ASC_READ_ERROR, NULL);
+		}
+		len = tcmu_iovec_length(iovec, iov_cnt);
+		iov.iov_base = malloc(len);
+		if (!iov.iov_base) {
+			tcmu_err("out of memory\n");
+			return tcmu_set_sense_data(sense, HARDWARE_ERROR,
+						   ASC_INTERNAL_TARGET_FAILURE,
+						   NULL);
+		}
+
+		iov.iov_len = len;
+
+		ret = read(dev, &iov, iov_cnt, offset);
+		if (ret != len) {
+			tcmu_err("read failed: %m\n");
+			free(iov.iov_base);
+			return tcmu_set_sense_data(sense, MEDIUM_ERROR,
+						   ASC_READ_ERROR, NULL);
+		}
+		cmp_offset = tcmu_compare_with_iovec(iov.iov_base,
+						     iovec,
+						     len);
+		if (cmp_offset != -1) {
+			tcmu_err("tcmu compare iovec fail\n");
+			rc = tcmu_set_sense_data(sense, MISCOMPARE,
+						 ASC_MISCOMPARE_DURING_VERIFY_OPERATION,
+						 &cmp_offset);
+		}
+
+		free(iov.iov_base);
+
+		if (rc != SAM_STAT_GOOD) {
+			tcmu_err("write_verify failed: %m\n");
+			return tcmu_set_sense_data(sense, MEDIUM_ERROR,
+						   ASC_READ_ERROR, NULL);
+
+		tcmu_seek_in_iovec(iovec, ret);
+		remaining -= ret;
+		}
+	}
```

Comment:
Something got messed up in one of your uploads. That bracket was supposed to be before the tcmu_seek_in_iovec call right?

}
tcmu_seek_in_iovec(iovec, ret); 
remaining -= ret;

Do not worry about re-uploading. I will fix it up when I merge.